audio-echo/app/src/main/java/com/google/sample/echo/MainActivity.java

```diff
@@ -120,11 +133,46 @@ private void queryNativeAudioParameters() {
         nativeSampleFormat ="";    //TODO: find a way to get the native audio format
     }
     private void updateNativeAudioUI() {
-        status_view.setText("nativeSampleRate    = " + nativeSampleRate + "\n" +
-                "nativeSampleBufSize = " + nativeSampleBufSize + "\n" +
-                "nativeSampleFormat  = " + nativeSampleFormat);
+        if(permissionGranted) {
+            status_view.setText("nativeSampleRate    = " + nativeSampleRate + "\n" +
+                    "nativeSampleBufSize = " + nativeSampleBufSize + "\n" +
+                    "nativeSampleFormat  = " + nativeSampleFormat);
+        } else {
+            status_view.setText("Error: Resource Permission not Granted" +
+                                "could not record Audio\n"    +
+                                 "Please restart App and grant the necessary permissions");
+        }
+    }
+    private void requestResourcePermission() {
+        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO) !=
+                PackageManager.PERMISSION_GRANTED) {
+            ActivityCompat.requestPermissions(this, new String[]{
+                            Manifest.permission.RECORD_AUDIO,
+                            Manifest.permission.MODIFY_AUDIO_SETTINGS,
+                            Manifest.permission.WRITE_EXTERNAL_STORAGE },
```

Comment:
But does the audio-echo app write to /sdcard/data?  (it seems that ENABLE_LOG is commented, so maybe the permission should also be guarded by a similar flag that you could query from C++).

What is AndroidLog advantage over logcat for logging text or time?